This should make conky compatible with the karmic kernel (2.6.31) and maintain
also compatibility with older kernels.
See LP #435571.

Index: conky-1.7.2/src/linux.c
===================================================================
--- conky-1.7.2.orig/src/linux.c	2009-08-08 22:05:55.000000000 +0200
+++ conky-1.7.2/src/linux.c	2009-10-15 18:33:30.000000000 +0200
@@ -858,23 +858,16 @@
 	}
 
 	if (strcmp(dir, "/sys/class/hwmon/") == 0) {
-		if (*buf) {
-			/* buf holds result from get_first_file_in_a_directory() above,
-			 * e.g. "hwmon0" -- append "/device" */
-			strcat(buf, "/device");
-		} else {
-			/* dev holds device number N as a string,
-			 * e.g. "0", -- convert to "hwmon0/device" */
-			sprintf(buf, "hwmon%s/device", dev);
-			dev = buf;
-		}
-	}
-
-	/* At least the acpitz hwmon doesn't have a 'device' subdir,
-	 * so check it's existence and strip it from buf otherwise. */
-	snprintf(path, 255, "%s%s", dir, dev);
-	if (stat(path, &st)) {
-		buf[strlen(buf) - 7] = 0;
+	  if (*buf) {
+	    /* buf holds result from get_first_file_in_a_directory() above,
+	     * e.g. "hwmon0" -- append "/device" */
+	    strcat(buf, "/device");
+	  } else {
+	    /* dev holds device number N as a string,
+	     * e.g. "0", -- convert to "hwmon0/device" */
+	    sprintf(buf, "hwmon%s/device", dev);
+	    dev = buf;
+	  }
 	}
 
 	/* change vol to in, tempf to temp */
@@ -884,16 +877,25 @@
 		type = "temp";
 	}
 
+	/* construct path */
 	snprintf(path, 255, "%s%s/%s%d_input", dir, dev, type, n);
-	strncpy(devtype, path, 255);
 
-	/* open file */
+	/* first, attempt to open file in /device */
 	fd = open(path, O_RDONLY);
 	if (fd < 0) {
-		CRIT_ERR(NULL, NULL, "can't open '%s': %s\nplease check your device or remove this "
-			"var from "PACKAGE_NAME, path, strerror(errno));
+
+	  /* if it fails, strip the /device from dev and attempt again */
+	  buf[strlen(buf) - 7] = 0;
+	  snprintf(path, 255, "%s%s/%s%d_input", dir, dev, type, n);
+	  fd = open(path, O_RDONLY);
+	  if (fd < 0) {
+	    CRIT_ERR(NULL, NULL, "can't open '%s': %s\nplease check your device or remove this "
+		     "var from "PACKAGE_NAME, path, strerror(errno));
+	  }
 	}
 
+	strncpy(devtype, path, 255);
+
 	if (strcmp(type, "in") == 0 || strcmp(type, "temp") == 0
 			|| strcmp(type, "tempf") == 0) {
 		*divisor = 1;
